lib/file-source/src/paths_provider/mod.rs

```diff
@@ -27,7 +27,9 @@ pub mod glob;
 pub trait PathsProvider {
```

Comment:
One of the design choices for the path provider was that it's never supposed to push it's error to the caller while enumerating the paths.

My thought process was this.
If the paths provider errors out - how would the caller - file server handle it? It is a meaningful error condition to recognize and act upon there? I deemed not, and decided that there's no good purpose to allow errors to be pushed to the file server. If the error is temporary - the valid behaviour would be to just log and return an empty iter. If the error is critial, and we should terminate the file server - then we should pass the error, however that situation just couldn't occur at the time.
There is another way I can think of that file server can react to an error - and that is, upon receiving a certain kind of error - just skip the current update, and keep using the paths obtained from the previous iterations. It may be useful for some kinds of use cases, like when the paths provider suddenly became unavailable, but we would rather use stale data and hope for the path provider to become available again at next call than terminate the file server. That,s however, also wasn't useful at the time.

The situation may have changed, and the change might be justified, but I'd like to validate that this way of handling error is what we're looking for.

Reply:
That for the rationale. Also don't have a clear cut answer to this, but my thoughts were the following:

> It is a meaningful error condition to recognize and act upon there?

This information can be encoded in/inferred from the associated `Error` type. In the case of the `Glob` `PathProvider` and `GlobError`, the caller can decide that e.g. `InvalidIncludePattern` is unrecoverable because it is a programmer error, whereas a `WalkError` can be temporary (due to the directory structure it is executed in), and ignored.

I would even argue that the caller has more information available if the error is recoverable or not, rather than deciding it in the `paths` implementation. With this change, the caller can still decide to ignore all errors, but the inverse is not true. If the `paths` implementation decides to unwrap and panic, the caller has little chance to override this behavior.

Reply:
> This information can be encoded in/inferred from the associated `Error` type.

Correct. The caller can't be coupled with an exact error type to carry this information though - otherwise it won't be generic. To establish a generic protocol we either have to add a trait to the error type, or use a trait-owned wrapper type to encode this data - i.e. `Result<..., PathsError<Self::Error>>`, where `enum PathsError<T> { UsePreviousPaths(T), Fatal(T) }` or sth.

Reply:
Ah, you're right. We don't specialize the `PathProvider` yet in the `FileServer::run` function.

> i.e. Result<..., PathsError<Self::Error>>, where enum PathsError<T> { UsePreviousPaths(T), Fatal(T) }.

It makes the trait a bit more verbose, but it does exactly solve the problem we are discussing. I'm leaning for it.